### DISCUSSION TITLE
Support for multiple system configurations?

### DISCUSSION THREAD

**Discussion by sommerper** [YaLTeR/niri#776] [Q&A] [answered]
Hi!

I'm sharing my dot files between multiple machines and now I'm trying to figure out if it is possible to have configurations based on the machine Niri is running on.

1. I have one machine with an ultra wide monitor and a huion monitor with tablet capabilities.
2. Then I have one with a 4k monitor only.
3. And one with a 2k monitor with touch input.

I'm wondering what the best way to have a shared config for all machines at once. If it's at all possible.

Thanks!

**Comment by YaLTeR** [comment]
Hey, unfortunately there are no config includes and stuff yet. But for monitor config at least, you can use monitor model+make+serial to configure them all, i.e. `output "Dell Inc. Dell S2716DG #ASOwvAqQj0Dd" {` (see `niri msg outputs`).

**Reply by sommerper** [reply]
Thanks! I think that might be enough for my current use case.

**Comment by sodiboo** [comment]
This is maybe unhelpful in that it doesn't really have anything to do with niri directly, but one way to do this is through something like `home-manager`, which uses the very flexible Nix expression language which among other things lets you share most of a config, and apply specializations for individual machines, by importing a common module and such. I use home-manager to manage my niri config through `niri-flake`, and [here's how that looks in the config](https://github.com/sodiboo/system/blob/276e0c2e48f2aede39d66f713ef2bcbbf299457d/niri.mod.nix#L414-L477). As you can see, most of my config is shared between my workstation (`sodium`) and laptop (`nitrogen`), but the preset widths, monitor settings, as well as which key is my compose key, are set individually per host.

I also use NixOS and home-manager to install niri, because they are tightly coupled with the Nix package manager. But it's not necessary to do so. A minimal config could just import `niri.homeModules.config` and set the config file, without ever installing niri from home-manager.

If the *only* thing you need is per-device monitor config, then niri absolutely does have that build in as YaLTeR said. But if you'd like to customize other parts per host, you'll need something more sophisticated to manage your dotfiles. `home-manager` is just *one* such sophisticated solution that i personally use to manage my dotfiles. You can just as well use some other form of templating and just have bash scripts to substitute parts of it when you deploy your dotfiles.

**Reply by sommerper** [reply]
I have in fact tried setting up a fairly complex nix system but eventually gave up because of the the language and it missing debugging features. Maybe it's a skill issue but I don't see myself going back. Thanks for the suggestion though - I hope it can help others!

**Comment by lpnh** [comment, marked as answer]
I think `chezmoi` could help you with this. Here's the relevant doc: https://www.chezmoi.io/user-guide/manage-machine-to-machine-differences/

**Reply by sommerper** [reply]
I didn't know about that feature in chezmoi. Thanks for the info!
